Wikipedia/Code/WMFAuthenticationManager.swift

```diff
@@ -127,8 +104,17 @@ public class WMFAuthenticationManager: NSObject {
      *  @param userAlreadyLoggedInHandler     The handler called if a user was found to already be logged in
      *  @param failure     The handler for any errors
      */
-    @objc public func loginWithSavedCredentials(success:@escaping WMFAccountLoginResultBlock, userAlreadyLoggedInHandler:@escaping WMFCurrentlyLoggedInUserBlock, failure:@escaping WMFErrorHandler){
+    public func loginWithSavedCredentials(_ loginURL: URL? = LoginSite.wikipedia.url, success: @escaping WMFAccountLoginResultBlock, userAlreadyLoggedInHandler: @escaping WMFCurrentlyLoggedInUserBlock, failure: @escaping WMFErrorHandler, completion: @escaping () -> Void = {}) {
         
+        defer {
+            completion()
```

Comment:
Is the intent that this be called after whichever other block (success, alreadyLoggedIn, failure) is called?

Reply:
yes

Reply:
The code inside `defer { }` will be executed when that function returns, it doesn’t take into account the various async calls. So as it is currently, the `completion` is called before  any of the other blocks are called

This might be a good opportunity to turn this into a method with a single completion block with a result type:

```
    enum LoginResult {
        case success(_: WMFAccountLoginResult)
        case alreadyLoggedIn(_: WMFCurrentlyLoggedInUser)
        case failure(_: Error)
    }

    ... completion: @escaping (LoginResult) -> Void ...
```

Reply:
Thanks, I will update the completion part. Given that `WMFAuthenticationManager` relies on `failure` and `success` blocks in many places (methods that are called inside this method), it seems like moving to a single completion would be a bigger change - could it be submitted as a follow-on?

Reply:
Yeah, fine to do as a follow on or not at all, just wanted to note it if it was an opportune time to consolidate